src/passes/MultiMemoryLowering.cpp

```diff
@@ -0,0 +1,394 @@
+          Abstract::getBinary(parent.pointerType, Abstract::Add),
+          builder.makeGlobalGet(global, parent.pointerType),
+          curr->ptr);
+      }
+    };
+    Replacer(*this, *wasm).run(getPassRunner(), wasm);
+  }
+
+  // Returns the global name for the given idx. There is no global for the first
+  // idx, so an empty name is returned
+  Name getOffsetGlobal(Index idx) {
+    // There is no offset global for the first memory
+    if (idx == 0) {
+      return Name();
+    }
+
+    // Since there is no offset global for the first memory, we need to
+    // subtract one when indexing into the offsetGlobalName vector
+    return offsetGlobalNames[idx - 1];
+  }
+
+  // Whether the idx represents the last memory. Since there is no offset global
+  // for the first memory, the last memory is represented by the size of
+  // offsetGlobalNames
+  bool isLastMemory(Index idx) { return idx == offsetGlobalNames.size(); }
+
+  void prepCombinedMemory() {
+    pointerType = wasm->memories[0]->indexType;
+    memoryInfo = pointerType == Type::i32 ? Builder::MemoryInfo::Memory32
+                                          : Builder::MemoryInfo::Memory64;
+    isShared = wasm->memories[0]->shared;
+    for (auto& memory : wasm->memories) {
+      // We are assuming that each memory is configured the same as the first
+      // and assert if any of the memories does not match this configuration
+      assert(memory->shared == isShared);
+      assert(memory->indexType == pointerType);
+
+      // Calculating the total initial and max page size for the combined memory
+      // by totaling the initial and max page sizes for the memories in the
+      // module
+      totalInitialPages = totalInitialPages + memory->initial;
+      if (memory->hasMax()) {
+        totalMaxPages = totalMaxPages + memory->max;
+      }
+    }
+    // Ensuring valid initial and max page sizes that do not exceed the number
+    // of pages addressable by the pointerType
+    Address maxSize =
+      pointerType == Type::i32 ? Memory::kMaxSize32 : Memory::kMaxSize64;
+    if (totalMaxPages > maxSize || totalMaxPages == 0) {
+      totalMaxPages = Memory::kUnlimitedSize;
+    }
+    if (totalInitialPages > totalMaxPages) {
+      totalInitialPages = totalMaxPages;
+    }
+
+    // Creating the combined memory name so we can reference the combined memory
+    // in subsequent instructions before it is added to the module
+    combinedMemory = Names::getValidMemoryName(*wasm, "combined_memory");
+  }
+
+  void addOffsetGlobals() {
+    auto addGlobal = [&](Name name, size_t offset) {
+      auto global =
+        Builder::makeGlobal(name,
+                            pointerType,
+                            Builder(*wasm).makeConst(
+                              pointerType == Type::i32
+                                ? Literal::makeFromInt32(offset, pointerType)
+                                : Literal::makeFromInt64(offset, pointerType)),
```

Comment:
```suggestion
                              Literal::makeFromInt64(offset, pointerType)),
```
This works out: we create something of the right type, and if the type is 32, the offset is truncated to 32 bits anyhow, either way. That is, `makeFromInt64` is safe to use to create 32-bit numbers.